conwayste/src/ui/chatbox.rs

```diff
@@ -0,0 +1,184 @@
+        let rect = Rect::new(30.0, 600.0, 300.0, 15.0*CHAT_DISPLAY_LIMIT);
+        Chatbox {
+            id: widget_id,
+            history_len: len,
+            color: Color::from(css::VIOLET),
+            messages: VecDeque::with_capacity(len),
+            dimensions: rect,
+            hover: false,
+            action: UIAction::EnterText,
+            font: font,
+        }
+    }
+
+    /// Adds a message to the chatbox
+    ///
+    /// # Arguments
+    /// * `ctx` - GGEZ context
+    /// * `msg` - New chat message
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use ui::Chatbox;
+    ///
+    /// fn new(ctx: &mut Context) -> GameResult<MainState> {
+    ///     let font = Rc::new(Font::default());
+    ///     let chatbox_rect = Rect::new(0.0, 0.0, chat_pane_rect.w, chat_pane_rect.h);
+    ///     let mut chatbox = Chatbox::new(WidgetID::InGamePane1Chatbox, font, 5);
+    ///     chatbox.set_size(chatbox_rect);
+    ///     chatbox.add_message(String::new("Player 1: This is a new chat message");
+    ///     chatbox.add_message(String::new("-- This is a Server broadcast message -- ");
+    ///     chatbox.draw(ctx)?;
+    /// }
+    /// ```
+    ///
+    pub fn add_message(&mut self, msg: String) -> GameResult<()> {
+        if self.messages.len() + 1 > self.history_len {
+            self.messages.pop_front();
+        }
+        // FIXME ggez0.5
+        let text = Text::new(msg);
+        self.messages.push_back(text);
+        Ok(())
+    }
+}
+
+
+impl Widget for Chatbox {
+    fn id(&self) -> WidgetID {
+        self.id
+    }
+
+    fn size(&self) -> Rect {
+        self.dimensions
+    }
+
+    fn set_size(&mut self, new_dims: Rect) {
+        self.dimensions = new_dims;
+    }
+
+    fn translate(&mut self, point: Vector2<f32>)
+    {
+        self.dimensions.translate(point);
+    }
+
+    fn on_hover(&mut self, point: &Point2<f32>) {
+        self.hover = within_widget(point, &self.dimensions);
+        //if self.hover {
+        //    println!("Hovering over Chatbox, \"{:?}\"", self.label.dimensions);
+        //}
+    }
+
+    fn on_click(&mut self, _point: &Point2<f32>) -> Option<(WidgetID, UIAction)>
+    {
+        let hover = self.hover;
+        self.hover = false;
+
+        if hover {
+            println!("Clicked within Chatbox");
+            return Some((self.id, self.action));
+        }
+
+        None
+    }
+
+    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
+        let origin = self.dimensions.point();
+
+        if self.hover {
+            // Add in a teal border while hovered. Color checkbox differently to indicate  hovered state.
+            let border_rect = Rect::new(self.dimensions.x-1.0, self.dimensions.y-1.0, self.dimensions.w + 4.0, self.dimensions.h + 4.0);
+            let hovered_border = graphics::Mesh::new_rectangle(ctx, DrawMode::stroke(2.0), border_rect, Color::from(css::TEAL))?;
```

Comment:
This color would be better as a constant. Maybe `ON_HOVER_COLOR` or that prefixed with `CHATBOX_`.

Reply:
It's already one of my tasks for the PR before we merge :smile: